File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Ios/IosVerifier.cs
using System.Security.Cryptography;
using System.Text;
using Azure.AI.Vision.Face.DeviceAttestation.Configuration;
using Azure.AI.Vision.Face.DeviceAttestation.Services;

namespace Azure.AI.Vision.Face.DeviceAttestation.Ios;

/// <summary>Result of a per-call App Attest assertion verification.</summary>
internal sealed class OngoingAssertionResult
{
    public bool Ok { get; init; }
    public string? Reason { get; init; }
    public string? Message { get; init; }
    public long SignCount { get; init; }
}

/// <summary>
/// iOS App Attest verifier — the seam the handlers depend on. Delegates the
/// attestation + assertion logic to <see cref="AppAttestVerification"/>.
/// </summary>
internal static class IosVerifier
{
    /// <summary>Verify an App Attest attestation (registration).</summary>
    public static Task<AuthVerificationResult> VerifyAsync(AttestationContext ctx, AttestationMessageData messageData, string attestJson)
        => Task.FromResult(AppAttestVerification.Verify(ctx.Config, ctx.Logger, messageData, attestJson));

    /// <summary>Expected rpIdHash: SHA-256 of the configured iOS App ID (lowercase hex).</summary>
    public static string? GetExpectedIosRpIdHash(AttestationConfig config)
    {
        if (string.IsNullOrEmpty(config.IosAppId))
        {
            return null;
        }
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(config.IosAppId))).ToLowerInvariant();
    }

    /// <summary>Verify a fresh per-call assertion against the persisted credCert.</summary>
    public static OngoingAssertionResult VerifyIosOngoingAssertion(
        string credCertPem,
        byte[] blob,
        string assertion,
        string expectedRpIdHash,
        long lastSignCount)
        => AppAttestVerification.VerifyOngoingAssertion(credCertPem, blob, assertion, expectedRpIdHash, lastSignCount);
}

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/ios/IosVerifier.java
package com.azure.ai.vision.face.deviceattestation.ios;

import java.nio.charset.StandardCharsets;

import com.azure.ai.vision.face.deviceattestation.AttestationContext;
import com.azure.ai.vision.face.deviceattestation.config.AttestationConfig;
import com.azure.ai.vision.face.deviceattestation.crypto.HashUtils;
import com.azure.ai.vision.face.deviceattestation.crypto.HexUtils;
import com.azure.ai.vision.face.deviceattestation.services.AttestationMessageData;
import com.azure.ai.vision.face.deviceattestation.services.AuthVerificationResult;

/**
 * iOS App Attest verifier — the seam the handlers depend on. Delegates the
 * attestation + assertion logic to {@link AppAttestVerification}.
 */
public final class IosVerifier {

    /** Verify an App Attest attestation (registration). */
    public static AuthVerificationResult verify(AttestationContext ctx, AttestationMessageData messageData, String attestJson) {
        return AppAttestVerification.verify(ctx.config, ctx.logger, messageData, attestJson);
    }

    /** Expected rpIdHash: SHA-256 of the configured iOS App ID (lowercase hex). */
    public static String getExpectedIosRpIdHash(AttestationConfig config) {
        if (config.iosAppId() == null || config.iosAppId().isEmpty()) {
            return null;
        }
        return HexUtils.toHex(HashUtils.sha256(config.iosAppId().getBytes(StandardCharsets.UTF_8)));
    }

    /** Verify a fresh per-call assertion against the persisted credCert. */
    public static OngoingAssertionResult verifyIosOngoingAssertion(
            String credCertPem,
            byte[] blob,
            String assertion,
            String expectedRpIdHash,
            long lastSignCount) {
        return AppAttestVerification.verifyOngoingAssertion(credCertPem, blob, assertion, expectedRpIdHash, lastSignCount);
    }

    private IosVerifier() {
    }
}
